src/stdune/string.ml

```diff
@@ -213,6 +214,13 @@ module Set = struct
       (Format.pp_print_list Format.pp_print_string
          ~pp_sep:(fun fmt () -> Format.fprintf fmt "@ "))
       (to_list t)
+
+  let to_dyn t =
+    let open Dyn in
+    Variant
+      ("Set"
```

Comment:
I thought it was weird that this was creating a "Variant" of variable arity. 
@diml proposes to add new variants `Dyn.Set` and `Dyn.Map` for this which seems good to me. 
Do you think we should do this? (this PR or later?)

Reply:
Yh, I feel like sets and maps are distinctive enough data-structures to justify having their own representation

Reply:
Sure, we can do it for Map/Set. What about hash table? Do we put that under Map? It seems like it will be a bit less readable this way but it doesn't matter to me

Reply:
I added such constructors. The literal notation is a matter of taste. Feel free to suggest something else.

Reply:
I was surprised to find Array and List as separate constructors. I'd expect them to be collapsed because they represent the same thing. (I guess the excuse is that they are printed differently by pp?)
Similarly for Map and Hashtbl. It could be called something neutral like "Dict", but Map is already supposed to be neutral, I think.

Reply:
The patch looks good to me.

Reply:
These nuances weren't considered much in `Dyn.t`. The use case is basically to avoid writing boilerplate for sexp & pretty printing. Feel free to tweak it as you see fit.